File: apps/reading/src/api/recordPromptHtmlNativeRecursiveTwinMoCompose.ts
/**
 * Workstation record→prompt model decision + HTML-native recursive twin MO (pure).
 *
 * Operator vision: recorded insights/questions recursively inform prompts with
 * model choice + budget projection, while the HTML-native recursive twin MO
 * write pack remains ready — never injects prompts, never persists records,
 * never PDF primary.
 *
 * record_persisted / prompts_injected always false.
 * pdf_view_authorized / pdf_primary always false.
 * twin_written / charge_executed always false.
 * production_router_verdict always REJECT; live_router_authorized always false.
 */

import {
  composeWorkstationRecordPromptModelDecision,
  type WorkstationRecordPromptModelDecisionCompose,
  type WorkstationRecordPromptModelDecisionInput,
} from "./workstationRecordPromptModelDecisionCompose";
import {
  composeHtmlNativeRecursiveTwinMoWritePack,
  type HtmlNativeRecursiveTwinMoWritePackCompose,
  type HtmlNativeRecursiveTwinMoWritePackInput,
} from "./htmlNativeRecursiveTwinMoWritePackCompose";

export interface RecordPromptHtmlNativeRecursiveTwinMoInput {
  record_prompt: Omit<
    WorkstationRecordPromptModelDecisionInput,
    "operator_ack"
  >;
  html_pack: Omit<HtmlNativeRecursiveTwinMoWritePackInput, "operator_ack">;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface RecordPromptHtmlNativeRecursiveTwinMoCompose {
  session_id: string;
  parent_asset_id: string;
  week_id: string;
  record_prompt: WorkstationRecordPromptModelDecisionCompose;
  html_pack: HtmlNativeRecursiveTwinMoWritePackCompose;
  pack_ready: boolean;
  record_persisted: false;
  prompts_injected: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  twin_written: false;
  charge_executed: false;
  live_execution_authorized: false;
  draft_written: false;
  analysis_written: false;
  merge_executed: false;
  inventory_mutated: false;
  live_dispatched: false;
  pack_dispatched: false;
  backlog_mutated: false;
  store_mutated: false;
  production_router_verdict: "REJECT";
  purchase_executed: false;
  notes: string[];
  authority: "record_prompt_html_native_recursive_twin_mo_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Record→prompt model decision overlay on HTML-native recursive twin MO pack.
 * Never injects prompts; never persists records; never PDF primary.
 */
export function composeRecordPromptHtmlNativeRecursiveTwinMo(
  input: RecordPromptHtmlNativeRecursiveTwinMoInput,
): RecordPromptHtmlNativeRecursiveTwinMoCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.record_prompt || typeof input.record_prompt !== "object") {
    throw new Error("record_prompt must be an object");
  }
  if (!input.html_pack || typeof input.html_pack !== "object") {
    throw new Error("html_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "record_persisted=false · prompts_injected=false · live_router_authorized=false",
    "pdf_view_authorized=false · pdf_primary=false",
    "twin_written=false · charge_executed=false",
    "production_router_verdict=REJECT",
  ];

  const record_prompt = composeWorkstationRecordPromptModelDecision({
    ...input.record_prompt,
    operator_ack: input.operator_ack,
  });
  notes.push(...record_prompt.notes.map((n) => `[record_prompt] ${n}`));

  const html_pack = composeHtmlNativeRecursiveTwinMoWritePack({
    ...input.html_pack,
    operator_ack: input.operator_ack,
  });
  notes.push(...html_pack.notes.map((n) => `[html_pack] ${n}`));

  const session_id = requireNonEmpty(record_prompt.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    record_prompt.parent_asset_id,
    "parent_asset_id",
  );
  const week_id = requireNonEmpty(html_pack.week_id, "week_id");

  const aligned =
    html_pack.session_id === session_id &&
    html_pack.parent_asset_id === parent_asset_id;
  if (!aligned) {
    notes.push(
      "session/parent mismatch between record_prompt and html_pack — pack_ready blocked",
    );
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      aligned &&
      record_prompt.pack_ready === true &&
      html_pack.pack_ready === true &&
      html_pack.production_router_verdict === "REJECT" &&
      record_prompt.prompts_injected === false &&
      html_pack.pdf_primary === false &&
      input.operator_ack === true;
  } else {
    pack_ready =
      aligned &&
      input.operator_ack === true &&
      html_pack.production_router_verdict === "REJECT" &&
      record_prompt.prompts_injected === false &&
      (record_prompt.pack_ready === true || html_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — record→prompt + HTML-native recursive twin MO ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — record_prompt, html_pack, alignment, or operator_ack gate open",
    );
  }

  if (
    record_prompt.record_persisted !== false ||
    record_prompt.prompts_injected !== false ||
    record_prompt.live_router_authorized !== false ||
    record_prompt.secrets_stored !== false ||
    record_prompt.live_meter_read !== false ||
    html_pack.pdf_view_authorized !== false ||
    html_pack.pdf_primary !== false ||
    html_pack.twin_written !== false ||
    html_pack.charge_executed !== false ||
    html_pack.production_router_verdict !== "REJECT" ||
    html_pack.live_router_authorized !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("twin_written=false");
  notes.push("charge_executed=false");
  notes.push("live_execution_authorized=false");
  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("merge_executed=false");
  notes.push("inventory_mutated=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("production_router_verdict=REJECT");
  notes.push("purchase_executed=false");

  return {
    session_id,
    parent_asset_id,
    week_id,
    record_prompt,
    html_pack,
    pack_ready,
    record_persisted: false,
    prompts_injected: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    twin_written: false,
    charge_executed: false,
    live_execution_authorized: false,
    draft_written: false,
    analysis_written: false,
    merge_executed: false,
    inventory_mutated: false,
    live_dispatched: false,
    pack_dispatched: false,
    backlog_mutated: false,
    store_mutated: false,
    production_router_verdict: "REJECT",
    purchase_executed: false,
    notes,
    authority: "record_prompt_html_native_recursive_twin_mo_compose_advisory",
  };
}

export function formatRecordPromptHtmlNativeRecursiveTwinMoSummary(
  c: RecordPromptHtmlNativeRecursiveTwinMoCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `record_prompt_ready=${c.record_prompt.pack_ready} · ` +
    `html_pack_ready=${c.html_pack.pack_ready} · ` +
    `would_exceed=${c.record_prompt.would_exceed} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `prompts_injected=false · pdf_primary=false · record_persisted=false`
  );
}

File: substrate/record_prompt_html_native_recursive_twin_mo_compose.py
"""Record→prompt model decision + HTML-native recursive twin MO pack (pure).

record_persisted / prompts_injected always False.
pdf_view_authorized / pdf_primary always False.
twin_written / charge_executed always False.
production_router_verdict always REJECT; live_router_authorized always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.html_native_recursive_twin_mo_write_pack_compose import (
    HtmlNativeRecursiveTwinMoWritePackCompose,
    HtmlNativeRecursiveTwinMoWritePackComposeError,
    compose_html_native_recursive_twin_mo_write_pack,
)
from substrate.workstation_record_prompt_model_decision_compose import (
    WorkstationRecordPromptModelDecisionCompose,
    WorkstationRecordPromptModelDecisionComposeError,
    compose_workstation_record_prompt_model_decision,
)


class RecordPromptHtmlNativeRecursiveTwinMoComposeError(ValueError):
    """Fail-closed validation for record→prompt + HTML recursive twin MO pack."""


@dataclass(frozen=True)
class RecordPromptHtmlNativeRecursiveTwinMoCompose:
    session_id: str
    parent_asset_id: str
    week_id: str
    record_prompt: WorkstationRecordPromptModelDecisionCompose
    html_pack: HtmlNativeRecursiveTwinMoWritePackCompose
    pack_ready: bool
    record_persisted: bool
    prompts_injected: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    twin_written: bool
    charge_executed: bool
    live_execution_authorized: bool
    draft_written: bool
    analysis_written: bool
    merge_executed: bool
    inventory_mutated: bool
    live_dispatched: bool
    pack_dispatched: bool
    backlog_mutated: bool
    store_mutated: bool
    production_router_verdict: str
    purchase_executed: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "week_id": self.week_id,
            "record_prompt": self.record_prompt.to_dict(),
            "html_pack": self.html_pack.to_dict(),
            "pack_ready": self.pack_ready,
            "record_persisted": False,
            "prompts_injected": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "twin_written": False,
            "charge_executed": False,
            "live_execution_authorized": False,
            "draft_written": False,
            "analysis_written": False,
            "merge_executed": False,
            "inventory_mutated": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "backlog_mutated": False,
            "store_mutated": False,
            "production_router_verdict": "REJECT",
            "purchase_executed": False,
            "notes": list(self.notes),
            "authority": (
                "record_prompt_html_native_recursive_twin_mo_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_record_prompt_html_native_recursive_twin_mo(
    *,
    record_prompt: object,
    html_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> RecordPromptHtmlNativeRecursiveTwinMoCompose:
    """Record→prompt + HTML recursive twin MO. Never injects or persists."""
    if not isinstance(operator_ack, bool):
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(record_prompt, dict):
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(
            "record_prompt must be an object"
        )
    if not isinstance(html_pack, dict):
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(
            "html_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "record_persisted=false · prompts_injected=false · live_router_authorized=false",
        "pdf_view_authorized=false · pdf_primary=false",
        "twin_written=false · charge_executed=false",
        "production_router_verdict=REJECT",
    ]

    try:
        rp = compose_workstation_record_prompt_model_decision(
            session_id=record_prompt.get("session_id"),
            parent_asset_id=record_prompt.get("parent_asset_id"),
            records=record_prompt.get("records"),
            user_prompt=record_prompt.get("user_prompt"),
            selected_model_id=record_prompt.get("selected_model_id"),
            models=record_prompt.get("models"),
            daily_cap_usd=record_prompt.get("daily_cap_usd"),
            spent_usd=record_prompt.get("spent_usd"),
            operator_ack=operator_ack,
            placement=record_prompt.get("placement"),
            max_context_lines=record_prompt.get("max_context_lines"),
            projected_cost_usd_high=record_prompt.get("projected_cost_usd_high"),
            projected_cost_usd_low=record_prompt.get("projected_cost_usd_low"),
            bench_bests=record_prompt.get("bench_bests"),
            focus_task=record_prompt.get("focus_task"),
            nd_shadow=record_prompt.get("nd_shadow"),
        )
    except WorkstationRecordPromptModelDecisionComposeError as e:
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(str(e)) from e
    notes.extend(f"[record_prompt] {n}" for n in rp.notes)

    try:
        hp = compose_html_native_recursive_twin_mo_write_pack(
            html_view=html_pack.get("html_view"),
            twin_mo=html_pack.get("twin_mo"),
            operator_ack=operator_ack,
            require_both=html_pack.get("require_both"),
        )
    except HtmlNativeRecursiveTwinMoWritePackComposeError as e:
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(str(e)) from e
    notes.extend(f"[html_pack] {n}" for n in hp.notes)

    session = _require_nonempty(rp.session_id, field="session_id")
    parent = _require_nonempty(rp.parent_asset_id, field="parent_asset_id")
    week = _require_nonempty(hp.week_id, field="week_id")

    aligned = hp.session_id == session and hp.parent_asset_id == parent
    if not aligned:
        notes.append(
            "session/parent mismatch between record_prompt and html_pack — pack_ready blocked"
        )

    if require:
        pack_ready = (
            aligned
            and rp.pack_ready is True
            and hp.pack_ready is True
            and hp.production_router_verdict == "REJECT"
            and rp.prompts_injected is False
            and hp.pdf_primary is False
            and operator_ack is True
        )
    else:
        pack_ready = (
            aligned
            and operator_ack is True
            and hp.production_router_verdict == "REJECT"
            and rp.prompts_injected is False
            and (rp.pack_ready is True or hp.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — record→prompt + HTML-native recursive twin MO ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — record_prompt, html_pack, alignment, or operator_ack gate open"
        )

    if (
        rp.record_persisted is not False
        or rp.prompts_injected is not False
        or rp.live_router_authorized is not False
        or rp.secrets_stored is not False
        or rp.live_meter_read is not False
        or hp.pdf_view_authorized is not False
        or hp.pdf_primary is not False
        or hp.twin_written is not False
        or hp.charge_executed is not False
        or hp.production_router_verdict != "REJECT"
        or hp.live_router_authorized is not False
    ):
        raise RecordPromptHtmlNativeRecursiveTwinMoComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "record_persisted=false",
            "prompts_injected=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "twin_written=false",
            "charge_executed=false",
            "live_execution_authorized=false",
            "draft_written=false",
            "analysis_written=false",
            "merge_executed=false",
            "inventory_mutated=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "backlog_mutated=false",
            "store_mutated=false",
            "production_router_verdict=REJECT",
            "purchase_executed=false",
        )
    )

    return RecordPromptHtmlNativeRecursiveTwinMoCompose(
        session_id=session,
        parent_asset_id=parent,
        week_id=week,
        record_prompt=rp,
        html_pack=hp,
        pack_ready=pack_ready,
        record_persisted=False,
        prompts_injected=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        twin_written=False,
        charge_executed=False,
        live_execution_authorized=False,
        draft_written=False,
        analysis_written=False,
        merge_executed=False,
        inventory_mutated=False,
        live_dispatched=False,
        pack_dispatched=False,
        backlog_mutated=False,
        store_mutated=False,
        production_router_verdict="REJECT",
        purchase_executed=False,
        notes=tuple(notes),
        authority="record_prompt_html_native_recursive_twin_mo_compose_advisory",
    )


def format_record_prompt_html_native_recursive_twin_mo_summary(
    c: RecordPromptHtmlNativeRecursiveTwinMoCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"record_prompt_ready={c.record_prompt.pack_ready} · "
        f"html_pack_ready={c.html_pack.pack_ready} · "
        f"would_exceed={c.record_prompt.would_exceed} · "
        f"verdict={c.production_router_verdict} · "
        f"prompts_injected=false · pdf_primary=false · record_persisted=false"
    )


__all__ = [
    "RecordPromptHtmlNativeRecursiveTwinMoCompose",
    "RecordPromptHtmlNativeRecursiveTwinMoComposeError",
    "compose_record_prompt_html_native_recursive_twin_mo",
    "format_record_prompt_html_native_recursive_twin_mo_summary",
]
